nil pointer dereference in `*Subscription.ConsumerInfo()` can cause deadlock

### Observed behavior

Deadlock or panic.

### Expected behavior

Neither deadlock nor panic, just error:

```
nats: invalid subscription type
exit status 1
```

### Server and client version

nats-server: `v2.12.2`
nats.go `v1.46.0` and later

### Host environment

Don't think it is relevant but `Fedora Linux 42`, `amd64`.

### Steps to reproduce

Hi there.

While experimenting with the latest release of the library on one of my work
projects, I encountered some strange behavior.

The following code deadlocks:

```go
package main

import (
	"log"

	"github.com/nats-io/nats.go"
)

func main() {
	log.SetFlags(0)

	conn, err := nats.Connect("nats://127.0.0.1:4222")
	if err != nil {
		log.Fatal(err)
	}
	defer conn.Close()

	sub, err := conn.Subscribe("test", func(_ *nats.Msg) {})
	if err != nil {
		log.Fatal(err)
	}
	defer sub.Unsubscribe()

	consumer, err := sub.ConsumerInfo()
	if err != nil {
		log.Fatal(err)
	}

	log.Printf("%#v", consumer)
}
```

If the code above is rewritten as:


```go
package main

import (
	"log"

	"github.com/nats-io/nats.go"
)

func main() {
	log.SetFlags(0)

	conn, err := nats.Connect("nats://127.0.0.1:4222")
	if err != nil {
		log.Fatal(err)
	}

	sub, err := conn.Subscribe("test", func(_ *nats.Msg) {})
	if err != nil {
		conn.Close()
		log.Fatal(err)
	}

	consumer, err := sub.ConsumerInfo()
	if err != nil {
		sub.Unsubscribe()
		conn.Close()
		log.Fatal(err)
	}

	log.Printf("%#v", consumer)

	sub.Unsubscribe()
	conn.Close()
}
```

the program no longer deadlocks, however, it crashes due to a nil dereference:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x68 pc=0x649776]

goroutine 1 [running]:
github.com/nats-io/nats%2ego.(*Subscription).ConsumerInfo(0x0?)
        github.com/nats-io/nats.go@v1.48.0/js.go:2839 +0xb6
main.main()
        example.org/nats-bug/nats-bug.go:23 +0x129
```

This happens because of the change made in commit 3d0a13a, which was included in [v1.46.0](https://github.com/nats-io/nats.go/releases/tag/v1.46.0) release.
On the line 2839, after checking that `sub.jsi` is `nil` or `sub.jsi.consumer == _EMPTY_`, method tries to retrieve the value of `sub.jsi.ordered`, which, causes a nil dereference because `sub.jsi` is nil in this case.

Since there is no deferred mutex unlock, any deferred call that requires lock on the same mutex will result in a deadlock.

--- a/js.go
+++ b/js.go
@@ -2835,12 +2835,15 @@ func ConsumerFilterSubjects(subjects ...string) SubOpt {
 func (sub *Subscription) ConsumerInfo() (*ConsumerInfo, error) {
 	sub.mu.Lock()
 	// TODO(dlc) - Better way to mark especially if we attach.
-	if sub.jsi == nil || sub.jsi.consumer == _EMPTY_ {
-		if sub.jsi.ordered {
-			sub.mu.Unlock()
+	if sub.jsi == nil {
+		sub.mu.Unlock()
+		return nil, ErrTypeSubscription
+	} else if sub.jsi.consumer == _EMPTY_ {
+		ordered := sub.jsi.ordered
+		sub.mu.Unlock()
+		if ordered {
 			return nil, ErrConsumerInfoOnOrderedReset
 		}
-		sub.mu.Unlock()
 		return nil, ErrTypeSubscription
 	}
 

--- a/test/nats_test.go
+++ b/test/nats_test.go
@@ -661,6 +661,27 @@ func TestBadSubjectsAndQueueNames(t *testing.T) {
 	}
 }
 
+func TestTypeSubscription(t *testing.T) {
+	s := RunServerOnPort(TEST_PORT)
+	defer s.Shutdown()
+
+	nc, err := nats.Connect(fmt.Sprintf("127.0.0.1:%d", TEST_PORT))
+	if err != nil {
+		t.Fatalf("Error connecting: %v", err)
+	}
+	defer nc.Close()
+
+	// Make sure that ConsumerInfo() returns invalid subscription type error
+	sub, err := nc.Subscribe("foo", func(_ *nats.Msg) {})
+	if err != nil {
+		t.Fatalf("Error subscribing: %v", err)
+	}
+
+	if _, err := sub.ConsumerInfo(); err != nats.ErrTypeSubscription {
+		t.Fatalf("Expected an error about invalid subscription type, got %v", err)
+	}
+}
+
 func BenchmarkNextMsgNoTimeout(b *testing.B) {
 	s := RunServerOnPort(TEST_PORT)
 	defer s.Shutdown()

